File: core/sds-aspect-meta-model-java/src/main/java/io/openmanufacturing/sds/metamodel/impl/DefaultStructuredValue.java
/*
 * Copyright (c) 2021 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for additional
 * information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package io.openmanufacturing.sds.metamodel.impl;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.StringJoiner;

import io.openmanufacturing.sds.metamodel.StructuredValue;
import io.openmanufacturing.sds.metamodel.Type;
import io.openmanufacturing.sds.metamodel.loader.MetaModelBaseAttributes;
import io.openmanufacturing.sds.metamodel.visitor.AspectVisitor;

public class DefaultStructuredValue extends DefaultCharacteristic implements StructuredValue {
   private final String deconstructionRule;
   private final List<Object> elements;

   public DefaultStructuredValue( final MetaModelBaseAttributes metaModelBaseAttributes,
         final Type dataType, final String deconstructionRule, final List<Object> elements ) {
      super( metaModelBaseAttributes, Optional.of( dataType ) );
      this.deconstructionRule = deconstructionRule;
      this.elements = elements;
   }

   @Override
   public String getDeconstructionRule() {
      return deconstructionRule;
   }

   @Override
   public List<Object> getElements() {
      return elements;
   }

   /**
    * Accepts an Aspect visitor
    *
    * @param visitor The visitor to accept
    * @param <T> The result type of the traversal operation
    * @param <C> The context of the visitor traversal
    */
   @Override
   public <T, C> T accept( final AspectVisitor<T, C> visitor, final C context ) {
      return visitor.visitStructuredValue( this, context );
   }

   @Override
   public String toString() {
      return new StringJoiner( ", ", DefaultStructuredValue.class.getSimpleName() + "[", "]" )
            .add( "deconstructionRule='" + deconstructionRule + "'" )
            .add( "elements=" + elements )
            .toString();
   }

   @Override
   public boolean equals( final Object o ) {
      if ( this == o ) {
         return true;
      }
      if ( o == null || getClass() != o.getClass() ) {
         return false;
      }
      if ( !super.equals( o ) ) {
         return false;
      }
      final DefaultStructuredValue that = (DefaultStructuredValue) o;
      return Objects.equals( deconstructionRule, that.deconstructionRule ) &&
            Objects.equals( elements, that.elements );
   }

   @Override
   public int hashCode() {
      return Objects.hash( super.hashCode(), deconstructionRule, elements );
   }
}

File: core/esmf-aspect-model-document-generators/src/test/resources/ts/snapshots/aspect-meta-model/characteristic/default-structured-value.ts
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for
 * additional information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

import {Characteristic, DefaultCharacteristic} from './default-characteristic';
import {Type} from '../type';
import {Property} from '../default-property';

export interface StructuredValue extends Characteristic {
    deconstructionRule: string;
    elements: Array<string | Property>;
}

export class DefaultStructuredValue extends DefaultCharacteristic implements StructuredValue {
    constructor(
        metaModelVersion: string,
        aspectModelUrn: string,
        name: string,
        private _deconstructionRule: string,
        private _elements: Array<string | Property>,
        dataType?: Type
    ) {
        super(metaModelVersion, aspectModelUrn, name, dataType);
    }

    public set deconstructionRule(value: string) {
        this._deconstructionRule = value;
    }

    public get deconstructionRule(): string {
        return this._deconstructionRule;
    }

    public set elements(value: Array<string | Property>) {
        this._elements = value;
    }

    public get elements(): Array<string | Property> {
        return this._elements;
    }
}
